src/sources/internal.rs

```diff
@@ -0,0 +1,178 @@
+use crate::{
+    event::metric::{Metric, MetricKind, MetricValue},
+    topology::config::{DataType, GlobalOptions, SourceConfig, SourceDescription},
+    Event,
+};
+use chrono::Utc;
+use futures::{
+    compat::Future01CompatExt,
+    future::{FutureExt, TryFutureExt},
+};
+use futures01::{sync::mpsc, Future, Sink, Stream};
+use metrics_core::Key;
+use metrics_runtime::{Controller, Measurement};
+use serde::{Deserialize, Serialize};
+use std::time::Duration;
+use stream_cancel::Tripwire;
+use tokio::timer::Interval;
+
+#[derive(Deserialize, Serialize, Debug, Clone, Default)]
+pub struct InternalConfig;
+
+inventory::submit! {
+    SourceDescription::new::<InternalConfig>("internal")
+}
+
+#[typetag::serde(name = "internal")]
+impl SourceConfig for InternalConfig {
+    fn build(
+        &self,
+        _name: &str,
+        _globals: &GlobalOptions,
+        out: mpsc::Sender<Event>,
+    ) -> crate::Result<super::Source> {
+        let (trigger, tripwire) = Tripwire::new();
+        trigger.disable(); // TODO: don't actually run forever
+        let fut = run(get_controller()?, out, tripwire).boxed().compat();
+        Ok(Box::new(fut))
+    }
+
+    fn output_type(&self) -> DataType {
+        DataType::Metric
+    }
+
+    fn source_type(&self) -> &'static str {
+        "internal"
+    }
+}
+
+fn get_controller() -> crate::Result<Controller> {
+    crate::metrics::CONTROLLER
+        .get()
+        .cloned()
+        .ok_or("metrics system not initialized".into())
+}
+
+async fn run(
+    controller: Controller,
+    mut out: mpsc::Sender<Event>,
+    mut tripwire: Tripwire,
+) -> Result<(), ()> {
+    let mut interval = Interval::new_interval(Duration::from_secs(2));
+    loop {
+        // Check for shutdown signal
+        if tripwire.poll().expect("polling tripwire").is_ready() {
+            return Ok(());
+        }
+
+        let (_, next) = interval.into_future().compat().await.expect("timers??");
```

Comment:
use `futures03::StreamExt::compat` and then use a `while let Some(_)`, timers will be set here. In tokio 0.2 the stream doesn't error it will just panic.